py-development.Rmd

```diff
@@ -0,0 +1,520 @@
+imagine that you open a file with the following lines of code within it.
+
+```{python, eval=FALSE}
+images = read_in_images(file_paths)
+gray_images = convert_to_grayscale(images)
+brightest_image = find_brightest_image(gray_images)
+```
+
+Just by looking at the function names, it is clear what this code is intended
+for and its main flow of operations is immediately visible. Inside each of
+these functions there might be 10-20 lines of code, so if we would not have
+modularized the code into separate functions with well chosen names, it
+would take longer to understand understand its overall purpose since there
+would be 30-60 lines of code to read instead of just three.
+
+### Creating functions {py-creating-functions}
+
+There are many useful functions already built into Python, and the ability to
+create your own allows you to string together any sequence of operations in
+ways that are tailored to your workflow. The `def` keyword lets us define a
+function with a name of our choice and an arbitrary number of input parameters.
+
+```{python}
+def sum_two_numbers(num1, num2):
+    return num1 + num2
+```
+
+This function accepts two input parameters, `num1` and `num2`, and returns
+their sum. Just as with variables names, function names are preferably written
+in `snake_case`, and avoid existing Python keyword and built-in names (a list of
+these is available [here][so-keywords-builtins], but instead of memorizing that
+list you can type the desired name into the Python interpreter to find out if
+it already exists).
+
+To execute the operations listed in the function, we can call the function
+and pass the two numbers we want to add as the arguments to the function.
+
+```{python}
+sum_two_numbers(2, 5)
+```
+
+The returned value can be assigned to a variable:
+
+```{python}
+number_sum = sum_two_numbers(2, 5)
+number_sum
+```
+
+A more versatile function could add any amount of numbers together and return
+their sum:
+
+```{python}
+def sum_all_numbers(list_of_numbers):
+    number_sum = 0
+    for number in list_of_numbers:
+        number_sum += number
+    return number_sum
+
+sum_all_numbers([1, 2, 3,])
+```
+
+A function can also return multiple outputs, e.g. we can return the number of
+elements in addition to their sum:
+
+```{python}
+def sum_and_len_all_numbers(list_of_numbers):
+    number_sum = 0
+    number_len = 0
+    for number in list_of_numbers:
+        number_sum += number
+        number_len += 1
+    return number_sum, number_len
+```
+
+To capture the output, we can either assign to a single name, a tuple, or
+assign both at the same time to different variables.
+
+```{python}
+sum_and_len_of_numbers = sum_and_len_all_numbers([1, 2, 3])
+sum_and_len_of_numbers
+```
+
+```{python}
+sum_of_numbers, len_of_numbers = sum_and_len_all_numbers([1, 2, 3])
+sum_of_numbers
+```
+
+```{python}
+len_of_numbers
+```
+
+> Note: When we defined the function `sum_two_numbers()`, we referred to `num1`
```

Comment:
This is a suggestion for how to include more advanced tips. We should probably make a nicer div (maybe bootstrap like green or blue color), but looking to get feedback on the idea before settling on design.

Reply:
Yea, not sure how we can do that across output document types (pdf vs ebook vs html). If you have a solid idea, let me know!

Also, I like this parahraph :+1:

Reply:
[This seems promising to me](https://stackoverflow.com/a/47294235/2166823), what do you think? We could create a function for "Advanced notes/Deep dives" (and maybe also "Exercises" if we want that to be more than a standard heading) that authors could call when inserting a similar paragraph as mine here.

Although this now appears technically feasible, and I like having a visually distinct element for advanced notes, it appears I am the only one who has written in this style so far (the rest follow your previous suggestion of having everything in paragraphs. So I am inclined to wander down the path of least resistance here and change my writing instead of encouraging everyone else to alter theirs.

Reply:
We can revisit this later. That link looks promising..